### PR TITLE
BUG: interpolate: RGI(nan) is nan

### LINKED ISSUE #6624: incorrect handling of nan by RegularGridInterpolator
RegularGridInterpolator does not seem to handle nans correctly when the interpolation method is nearest-neighbor. Sample script follows below.

--Bram
### 

import numpy as np
from scipy.interpolate import RegularGridInterpolator

f = RegularGridInterpolator(([1,2,3],), [10,20,30], bounds_error=False, fill_value=-999, method='linear')

print f([1]), f([0]), f([np.nan])
'# [ 10.] [-999.] [ nan]: correct handling'

f = RegularGridInterpolator(([1,2,3],), [10,20,30], bounds_error=False, fill_value=-999, method='nearest')

print f([1]), f([0]), f([np.nan])
'# [ 10.] [-999.] [ 30.] --> last value should be nan I guess...'

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

 closes gh-6624

#### What does this implement/fix?

Make RegularGridInterpolator have  `f(nan) = nan`.

#### Additional information

Will need to be rebased on gh-16070 first.

### CODE REVIEW

File: scipy/interpolate/_interpolate.py
Comment (mdhaber) [inline scipy/interpolate/_interpolate.py:2538]: Just a thought, in case the `axis` argument of `np.any` didn't come to mind.
```suggestion
        nans = np.any(np.isnan(xi), axis=tuple(range(1, xi.ndim)))
```
Then again, `tuple(range(...))` is a bit clumsy, so I can see why you still might prefer the comprehension.

I can't tell from the code whether `xi` is always 2D? Sometimes (e.g.`xi.T`) it looks like it, elsewhere (e.g. `xi.reshape(-1, xi_shape[-1])` not so much...)  In that case, this could be simplified to:

```suggestion
        nans = np.any(np.isnan(xi), axis=-1)
```

If it is 1D or 2D, seems like an `np.atleast_2d` at the beginning could simplify things.

File: scipy/interpolate/tests/test_interpolate.py
Comment (mdhaber) [inline scipy/interpolate/tests/test_interpolate.py:2749]:
```suggestion
        f = RegularGridInterpolator(([1, 2, 3],), [10, 20, 30], fill_value=1,
```
to show that this isn't just passing because the default fill values is `np.nan`.

**Review by mdhaber — APPROVED** [review]
Looks good; just minor thoughts. I'll merge either way; just LMK what you think about them.

File: scipy/interpolate/tests/test_interpolate.py
Comment (ev-br) [reply scipy/interpolate/tests/test_interpolate.py]: Thanks! There is a test with nans and non-nans below, but it wouldn't hurt to test it more.

File: scipy/interpolate/_interpolate.py
Comment (ev-br) [reply scipy/interpolate/_interpolate.py:2538]: I *think* `xi` is 2D at this point after the weird-looking reshape above. So your `(..., axis=-1)`suggestion should work, and if it makes the intent clearer to you, great, let's do it like this.

**Comment by ev-br** [comment]
BTW, there is probably a similar issue in np.searchsorted bases kinds of `interp1d`. May also fix them while we're at it. Not in this PR though.  Or even wait until there is an actual user report.

File: scipy/interpolate/_interpolate.py
Comment (mdhaber) [reply scipy/interpolate/_interpolate.py]: I think they are equally clear. Not a strong preference; I just tend toward vectorization over explicit loops unless there's a reason not to.

File: scipy/interpolate/tests/test_interpolate.py
Comment (mdhaber) [reply scipy/interpolate/tests/test_interpolate.py]: Yeah I saw. It's just a reflex from working with stats, as there have been a lot of issues with nans in variable numbers of dimensions.

**Comment by mdhaber** [comment]
> Not in this PR though.

That's fine. Up to you when to do it.

**Comment by mdhaber** [comment]
> but it wouldn't hurt to test it more.

OK, I went for it in ev-br/scipy#8, which also fixes the PEP8 issue. I didn't fix the wheel_optimized_gcc6 failure. I thought I'd leave that up to you, since you mentioned that you didn't like `with np.errstate`.

**Comment by mdhaber** [comment]
@ev-br there was another place in the 2D test where the warning was generated, so I silenced that. I'll merge when tests pass?

**Comment by mdhaber** [comment]
Thanks @ev-br!

I saw the note:
> Out-of-bounds checks, `out_of_bounds += x < grid[0]`, generate numpy warnings if `x` contains nans. These warnings should
propagate to user (since `x` is user input). Thus simply filter them out in the test suite.

With `bounds_error=True`, I agree that there should be a warning, but I'm not sure that about with `bounds_error=False`. If `nan` input will always produce `nan` output, then the fact that an `out_of_bounds` calculation is performed with `nan` input is an implementation detail, no? If so, I don't think a warning is necessary, since the comparison is unnecessary and has no effect on the output.

**Comment by ev-br** [comment]
Exactly. 

These checks are guarded by `if not self.bounds_error:` clause, so I believe it behaves just like you said. And I think there is not much more we should do about it _in the library code_. If a user feeds in nans, they may or may not get a warning, depending on implementation details.  There is no signalling / quiet nan distinction in numpy, for better or worse.

**Pull request merged**
